### PR TITLE
Add isValid() to RealmList, RealmResults, RealmQuery

### PR DESCRIPTION
**Pull request by zaki50**
refs #1431

TODO: 
- [x] for RealmList detecting that container object is alive or not
- [x] add `RealmQuery#isValid()`

### CODE REVIEW

**Comment by zaki50** [comment]
@realm/java review this please

File: realm/src/androidTest/java/io/realm/RealmQueryTest.java [hunk 1]
```diff
@@ -1369,4 +1369,69 @@ public void run() {
...
+
+    public void testIsValidOfLinkViewQuery() {
+        populateTestRealm(1);
+        final RealmList<Dog> list = testRealm.where(AllTypes.class).findFirst().getColumnRealmList();
+        final RealmQuery<Dog> query = list.where();
+        final long listLength = query.count();
+        assertTrue(query.isValid());
+
+        testRealm.beginTransaction();
+        {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmQueryTest.java]: C++ bracers

File: realm/src/androidTest/java/io/realm/RealmListTest.java [hunk 2]
```diff
@@ -294,6 +291,34 @@ public void testMoveUp_nonManagedMode() {
...
+        testRealm.close();
+        assertFalse(dogs.isValid());
+    }
+
+    public void testIsValidWhenParentRemoved() {
+        Owner owner = testRealm.where(Owner.class).findFirst();
+        RealmList<Dog> dogs = owner.getDogs();
+
+        testRealm.beginTransaction();
+        {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmListTest.java]: C++ bracers

File: realm/src/androidTest/java/io/realm/RealmQueryTest.java [hunk 3]
```diff
@@ -1369,4 +1369,69 @@ public void run() {
...
+    }
+
+    public void testIsValidOfRemovedParent() {
+        populateTestRealm(1);
+        final AllTypes obj = testRealm.where(AllTypes.class).findFirst();
+        final RealmQuery<Dog> query = obj.getColumnRealmList().where();
+        assertTrue(query.isValid());
+
+        testRealm.beginTransaction();
+        {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmQueryTest.java]: C++ parent

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 4]
```diff
@@ -173,6 +173,23 @@ private boolean containsDot(String s) {
     }
 
     /**
+     * Check if {@link io.realm.RealmQuery} is still valid to use ie. the {@link io.realm.Realm}
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: ie. -> i.e.

File: realm/src/main/java/io/realm/RealmQuery.java [hunk 5]
```diff
@@ -173,6 +173,23 @@ private boolean containsDot(String s) {
...
+     * instance hasn't been closed and base {@link io.realm.RealmResults} is still valid if exists.
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmQuery.java]: and any parent {@link io.realm.RealmResults} is still valid.

File: realm/src/main/java/io/realm/RealmResults.java [hunk 6]
```diff
@@ -87,6 +87,16 @@ TableOrView getTable() {
     }
 
     /**
+     * Check if {@link io.realm.RealmResults} is still valid to use ie. the {@link io.realm.Realm}
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmResults.java]: ie. -> i.e.
. after closed

**Comment by cmelchior** [comment]
Some small unit test and doc corrections, otherwise it looks good.

**Comment by zaki50** [comment]
@cmelchior all done.

File: realm/src/main/java/io/realm/RealmList.java [hunk 7]
```diff
@@ -98,6 +98,22 @@ public RealmList(E... objects) {
         this.realm = realm;
     }
 
+    /**
+     * Check if {@link io.realm.RealmResults} is still valid to use i.e. the {@link io.realm.Realm}
+     * instance hasn't been closed.
+     *
+     * @return {@code true} if still valid to use, {@code false} otherwise or if it is a
+     * standalone list object.
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmList.java]: remove list

File: realm/src/main/java/io/realm/RealmResults.java [hunk 8]
```diff
@@ -87,6 +87,16 @@ TableOrView getTable() {
     }
 
     /**
+     * Check if {@link io.realm.RealmResults} is still valid to use i.e. the {@link io.realm.Realm}
+     * instance hasn't been closed
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/RealmResults.java]: Missing .

**Comment by cmelchior** [comment]
Apart from two minor javadoc things :+1:

**Comment by zaki50** [comment]
after another +1, I'll squash and merge. anyone? @realm/java

**Comment by kneth** [comment]
:+1:

**Pull request merged**